nixos/modules/services/misc/conduit.nix

```diff
@@ -0,0 +1,118 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.conduit;
+
+  format = pkgs.formats.toml {};
+  namespacedSettings.global = cfg.settings;
+  configFile = format.generate "conduit.toml" namespacedSettings;
+in
+  {
+    options.services.conduit = {
+      enable = mkEnableOption "conduit";
+      iHaveReadTheFederationWarning = mkOption {
+        type = types.bool;
+        default = false;
+        example = true;
+        description = ''
+          Upstream documentation contains the following warning:
+          You should not join Matrix rooms without asking the admins first.
+          We do not know whether Conduit is safe for general use yet, so you
+          should assume there is some chance that it breaks rooms permanently
+          for all participating users. We are not aware of such a bug today,
+          but we would like to do more testing.
+
+          This option will be removed upon removal of this warning.
+        '';
+      };
+
+      package = mkOption {
+        type = types.package;
+        default = pkgs.conduit;
+        defaultText = "pkgs.conduit";
+        example = "pkgs.conduit";
+        description = ''
+          Package of the conduit matrix server to use.
+        '';
+      };
+
+      settings = mkOption {
+        type = format.type;
+        default = {
+        };
+        example = literalExample ''
+          {
+          server_name = "example.com";
+          port = 6167;
+          max_request_size = "20000000";
+          allow_registration = false;
+          allow_encryption = false;
+          allow_federation = false;
+          allow_jaeger = false;
+          trusted_servers = [ "matrix.org" ];
+          address = "127.0.0.1";
+          cache_capacity = 1073741824
+          max_concurrent_requests = 100
+          log = "info,state_res=warn,rocket=off,_=off,sled=off"
+          workers = 4
+          };
+        '';
+        description = ''
+          generates the conduit.toml configuration file. Refer to
+          <link xlink:href="https://gitlab.com/famedly/conduit/-/blob/master/conduit-example.toml"/>
+          for details on supported values.
+          Note that database_path should not be set because the services reliance on systemd StateDir.
+        '';
+      };
+    };
+
+
+    config = mkIf cfg.enable {
+      services.conduit.settings = {
+        database_path = "/var/lib/conduit/conduit.db";
+        port = mkDefault 6167;
+        max_request_size = mkDefault 20000000;
+        allow_registration = mkDefault false;
+        allow_encryption = mkDefault false;
+        allow_federation = mkDefault false;
+        allow_jaeger = mkDefault false;
+        trusted_servers = mkDefault [ "matrix.org" ];
+        address = mkDefault "127.0.0.1";
+      };
```

Comment:
Use freeFormType on the settings option and define these option defaults there instead.

Example: https://github.com/NixOS/nixpkgs/blob/84f649f693575eed9dedd9f05fde2a32b9d1e325/nixos/modules/services/audio/botamusique.nix